virtcontainers/netns_watcher/netns_watcher.c

```diff
@@ -0,0 +1,1251 @@
+		if (ret) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * Initialize global list of interfaces.
+ */
+void init_iface_list()
+{
+	int i;
+
+	for (i = 0; i < MAX_IFACES; i++) {
+		struct iface nif = {
+			.idx      = INIT_IF_IDX,
+			.hw_addr  = NULL,
+			.name     = NULL,
+			.mtu      = NULL,
+			.ip_addrs = NULL,
+		};
+
+		iface_list[i] = nif;
+	}
+}
+
+/*
+ * Fill a new interface from the information given by ifaddrs.
+ */
+int create_iface_from_ifaddrs(struct ifaddrs *ifa, struct iface *nif)
+{
+	int sock;
+	int family;
+	int ret = 0;
+	struct ifreq ifr;
+
+	if (!strncpy(ifr.ifr_name , nif->name, IFNAMSIZ-1)) {
+		printf("Failed to copy interface name %s", nif->name);
+		return -1;
+	}
+
+	sock = socket(AF_INET, SOCK_DGRAM, 0);
+	if (sock == -1) {
+		printf("Failed to open socket: %s", strerror(errno));
+		return -1;
+	}
+
+	if ((ret = ioctl(sock, SIOCGIFHWADDR, &ifr)) == -1) {
+		printf("Failed to get HW ADDR: %s", strerror(errno));
+		goto exit;
+	}
+
+	nif->hw_addr = hw_addr_to_string((unsigned char *)
+			ifr.ifr_hwaddr.sa_data);
+
+	if ((ret = ioctl(sock, SIOCGIFMTU, &ifr)) == -1) {
+		printf("Failed to get MTU: %s", strerror(errno));
+		goto exit;
+	}
+
+	if (asprintf(&(nif->mtu), "%d", ifr.ifr_mtu) == -1) {
+		printf("Failed allocating memory: %s\n",
+				strerror(errno));
+		goto exit;
+	}
+
+exit:
+	close(sock);
+	return ret;
+}
+
+/*
+ * Fill a new interface from the information given by ifaddrs.
+ */
+int add_ip_addr_to_iface_list(struct ifaddrs *ifa, unsigned int idx)
+{
+	unsigned char family;
+	int s;
+	char host[NI_MAXHOST];
+
+	family = ifa->ifa_addr->sa_family;
+	if (family != AF_INET && family != AF_INET6) {
+		return 0;
+	}
+
+	s = getnameinfo(ifa->ifa_addr,
+			(family == AF_INET) ? sizeof(struct sockaddr_in) :
+			sizeof(struct sockaddr_in6),
+			host, NI_MAXHOST, NULL, 0, NI_NUMERICHOST);
+	if (s) {
+		printf("Failed getnameinfo(): %s\n", gai_strerror(s));
+		return -1;
+	}
+
+	insert_ip_addr(idx, host, family);
+
+	return 0;
+}
+
+/* 
+ * Scan the network and store it into a global list.
+ *
+ * A precise description of the network is needed to make sure events received
+ * through the netlink socket will be properly interpreted.
+ */
+int scan_netns()
+{
+	struct ifaddrs *ifaddr, *ifa;
+	int n;
+	int ret = 0;
+
+	if (getifaddrs(&ifaddr) == -1) {
+		printf("Failed getifaddrs(): %s\n", strerror(errno));
+		return -1;
+	}
+
+	for (ifa = ifaddr, n = 0; ifa != NULL; ifa = ifa->ifa_next, n++) {
+		struct iface nif;
+		unsigned int if_idx;
+
+		if (!ifa->ifa_name) {
+			continue;
+		}
+
+		if_idx = if_nametoindex(ifa->ifa_name);
+		if (!if_idx) {
+			printf("Failed if_nametoindex() for %s: %s\n",
+					ifa->ifa_name, strerror(errno));
+			goto exit;
+		}
+
+		if (iface_list[if_idx].idx == INIT_IF_IDX) {
+			nif.name = strdup(ifa->ifa_name);
+
+			ret = create_iface_from_ifaddrs(ifa, &nif);
+			if (ret) {
+				free_iface(&nif);
+				goto exit;
+			}
+
+			nif.idx = if_idx;
+
+			if (if_idx >= MAX_IFACES) {
```

Comment:
I think MAX_IFACES=50 is far from enough.

if_idx is link index, if host has more than 50 interfaces, then you create a new container, it's interface link index is easily larger than 50.
So I think link index shouldn't be used as the array's index, the array index should be standalone counter.

Reply:
And I get a `segment fault`, to reproduce the error, you can create 50 containers, then try to exec

`$ netns_watcher  -d -n /proc/<pid>/ns/net -p hello -r /usr/local/bin/kata-runtime`

Pid should be process id of last container, and its interface index is larger than 50.

Reply:
About the table of interfaces, that's exactly what I was mentioning when I said I'd like to replace this with a hash table ;). I will make the change.

Now about the segmentation fault, I am not sure, and I'll try to reproduce !